riot/riot.py

```diff
@@ -62,60 +66,157 @@ def to_list(x: t.Union[_K, t.List[_K]]) -> t.List[_K]:
+            [self._path, "--version"], stderr=subprocess.STDOUT
+        )
+        version = Version(output.decode().strip().split(" ")[1])
+        self._version = version
 
-    def __post_init__(self, hint: _T_hint) -> None:
-        """Normalize the data."""
-        self._hint = str(hint)
+    def __hash__(self) -> int:
+        """Return the hash of this interpreter."""
+        return hash(self._path)
+
+    def __repr__(self) -> str:
+        """Return the repr containing the path and version of this interpreter."""
+        return f"{self.__class__.__name__}('{self.version}', '{self._path}')"
 
     def __str__(self) -> str:
-        """Return the path of the interpreter executable."""
+        """Return the repr of this interpreter."""
         return repr(self)
 
+    @staticmethod
+    def _matches(v1: Version, v2: Version) -> bool:
+        """Return if v2 matches v1.
+
+        >>> Interpreter._matches(Version("3.8.2"), Version("3"))
+        True
+        >>> Interpreter._matches(Version("3.8.2"), Version("3.8"))
+        True
+        >>> Interpreter._matches(Version("3.8.2"), Version("3.8.2"))
+        True
+        >>> Interpreter._matches(Version("3.8.2"), Version("3.8.3"))
+        False
+        >>> Interpreter._matches(Version("3.8.2"), Version("3.6"))
+        False
+        >>> Interpreter._matches(Version("3.8.2"), Version("3.7"))
+        False
+        >>> Interpreter._matches(Version("3.8.2"), Version("2"))
+        False
+        >>> Interpreter._matches(Version("3.8.2-dev"), Version("3.8.2"))
+        True
+        >>> Interpreter._matches(Version("3.8.2-alpha"), Version("3.8.2"))
+        True
+
+        FIXME: unfortunately Version will set undefined segments to 0 so
+        we cannot distinguish between "3.0.0" and "3".
+        >>> Interpreter._matches(Version("3.8.2"), Version("3.0.0"))
+        True
+        >>> Interpreter._matches(Version("3.8.2"), Version("3.8.0"))
+        True
+        """
+        if v2 == v1:
+            return True
+        elif v2.micro != v1.micro and v2.micro != 0:
+            return False
+        elif v2.minor != v1.minor and v2.minor != 0:
+            return False
+        elif v2.major != v1.major and v2.major != 0:
+            return False
+        return True
+
+    @classmethod
     @functools.lru_cache()
-    def version(self) -> str:
-        path = self.path()
+    def find(cls) -> t.List["Interpreter"]:
+        """Find all Python interpreters discoverable on the PATH."""
+        ex_dirs = os.environ.get("PATH", "").split(":")
+        # Use a list because order matters (need the first match).
+        interp_paths: t.List[str] = []
+        # packaging says that the regular expression needs to be compiled
+        # with the following flags
+        expr = re.compile(f"python{VERSION_PATTERN}$", re.IGNORECASE | re.VERBOSE)
+        for d in ex_dirs:
+            if not os.path.isdir(d):
+                continue
+            for f in os.listdir(d):
```

Comment:
As a rule of thumb, never do that. The pattern of calling `stat` before doing an I/O is at best unoptimized, at worst a security issue.

You should always do a single file operation and catch errors, like this:

```suggestion
        try:
            for f in os.listdir(d):
             [...]
        except NotADirectoryError:
             continue
```